### PR TITLE
Upgrade Alpine from 3.13 to 3.15

### PR DESCRIPTION
**Pull request by galibozek**
Upgrade Alpine from 3.13 to 3.15

## :warning: BREAKING :warning: 

Users with docker version < 20.10.6 have to upgrade their docker.

See: https://wiki.alpinelinux.org/wiki/Release_Notes_for_Alpine_3.14.0 & https://about.gitlab.com/blog/2021/08/26/its-time-to-upgrade-docker-engine/

### CODE REVIEW

**Comment by zeripath** [comment]
Have alpine sorted the reason why we had to downgrade back to 3.13 last year?

If not have sufficient numbers of people upgraded their dockers to prevent it from being problem?

**Comment by silverwind** [comment]
This should help with the issue described in https://github.com/go-gitea/gitea/pull/17994 related to old npm.

**Comment by zeripath** [comment]
Let me be clearer:

Alpine 3.14+ will cause everyone using gitea with docker less than version 20.10.6 to break.

We upgraded to alpine 3.14 and had multiple bug reports like:


https://about.gitlab.com/blog/2021/08/26/its-time-to-upgrade-docker-engine/

**Comment by zeripath** [comment]
So the question is has enough of the docker community moved over to update their docker installs and have they been bitten elsewhere yet?

One option is to do this but also add a check in to Gitea that will check if this bug is present and report that we cannot run.

**Comment by wxiaoguang** [comment]
That's a lot of new information for me, too.

If we can write the background in comment, maybe new developers won't have questions anymore.

**Comment by zeripath** [comment]
So looking at:

https://github.com/alpinelinux/docker-alpine/issues/156#issuecomment-997569074 & https://github.com/alpinelinux/docker-alpine/issues/156#issuecomment-931321798

May provide techniques for us to test at time of docker run if things are going to be broken.

```
if [ ! -x /bin/bash ]; then echo "I should not get printed!"; fi
```

```
Directory test also failed ( test -x /tmp then echo $? , got 1 )
```

**Comment by silverwind** [comment]
Docker is just a dependency in this case. We should document its minimum version and possible refuse to run if it's too old.

**Comment by eeyrjmr** [comment]
> Docker is just a dependency in this case. We should document its minimum version and possible refuse to run if it's too old. Question is, can a container detect the version of the docker daemon it runs on?

this might do it 
`cat` /proc/self/cgroup | grep -o  -e "docker-.*.scope" | head -n 1 | sed "s/docker-\(.*\).scope/\\1/"

File: docker/rootless/usr/local/bin/docker-entrypoint.sh [hunk 1]
```diff
@@ -1,5 +1,11 @@
 #!/bin/sh
 
+# Protect against buggy runc in docker <20.10.6 causing problems in with Alpine >= 3.14
+if [ ! -x /bin/sh ]; then
+  echo "Executable test for /bin/sh failed. Your Docker version is too old to run Alpine 3.+ and Gitea. You must upgrade Docker.";
```
Comment (zeripath) [inline docker/rootless/usr/local/bin/docker-entrypoint.sh]:
```suggestion
  echo "Executable test for /bin/sh failed. Your Docker version is too old to run Alpine 3.14+ and Gitea. You must upgrade Docker.";
```

**Review by zeripath — APPROVED** [review]

**Pull request merged**